framework/aster-frame/src/arch/x86/iommu/second_stage.rs

```diff
@@ -42,117 +67,71 @@ bitflags::bitflags! {
 #[repr(C)]
 pub struct PageTableEntry(u64);
 
-impl PageTableFlagsTrait for PageTableFlags {
-    fn new() -> Self {
-        Self::empty()
-    }
-
-    fn set_present(self, present: bool) -> Self {
-        self
-    }
-
-    fn set_writable(mut self, writable: bool) -> Self {
-        self.set(Self::WRITABLE, writable);
-        self
-    }
-
-    fn set_readable(mut self, readable: bool) -> Self {
-        self.set(Self::READABLE, readable);
-        self
-    }
-
-    fn set_accessible_by_user(self, accessible: bool) -> Self {
-        self
-    }
-
-    fn set_executable(self, executable: bool) -> Self {
-        self
-    }
-
-    fn is_present(&self) -> bool {
-        self.contains(Self::WRITABLE) || self.contains(Self::READABLE)
-    }
-
-    fn writable(&self) -> bool {
-        self.contains(Self::WRITABLE)
-    }
-
-    fn readable(&self) -> bool {
-        self.contains(Self::READABLE)
-    }
-
-    fn executable(&self) -> bool {
-        true
-    }
-
-    fn has_accessed(&self) -> bool {
-        self.contains(Self::ACCESSED)
-    }
-
-    fn is_dirty(&self) -> bool {
-        self.contains(Self::DIRTY)
-    }
-
-    fn accessible_by_user(&self) -> bool {
-        true
-    }
-
-    fn union(&self, other: &Self) -> Self {
-        (*self).union(*other)
-    }
-
-    fn remove(&mut self, flags: &Self) {
-        self.remove(*flags)
-    }
-
-    fn insert(&mut self, flags: &Self) {
-        self.insert(*flags)
-    }
-
-    fn is_huge(&self) -> bool {
-        // FIXME: this is super bad
-        false
-    }
-
-    fn set_huge(self, huge: bool) -> Self {
-        // FIXME: this is super bad
-        self
-    }
-}
-
 impl PageTableEntry {
     const PHYS_MASK: usize = 0xFFFF_FFFF_F000;
 }
 
 impl PageTableEntryTrait for PageTableEntry {
```

Comment:
This problem might be outside the scope of this PR, but I'd like to raise it here.

The following definition of `PageTableEntry`, where a `PageTableEntry` directly references a entry in the page table, violates the Rust's memory model.
```rust
#[repr(C)]
pub struct PageTableEntry(usize);
```

 - By Rust's memory mode, whenever someone creates a `&PageTableEntry`, the entry must be immutable. However, the accessed bit and the dirty bit of the entry might be set by the CPU at any time.
 - Similarly, whenever someone creates a `&mut PageTableEntry`, they claim the exclusive access to the entry. However, the accessed bit and the dirty bit may be modified by the CPU at any time.

Generally, I would suggest that we should use only [`ptr::read()`](https://doc.rust-lang.org/std/ptr/fn.read.html) and [`ptr::write()`](https://doc.rust-lang.org/std/ptr/fn.write.html) to access the page table entries. It's okay to consider this issue out of scope, but it has a lot of implications for how you design and implement other page table types, such as the cursors.